File: python/simple-linked-list/simple_linked_list.py
class Node(object):
    def __init__(self, value):
        self.v = value
        self.n = None

    def value(self):
        return self.v

    def next(self):
        return self.n


class LinkedList(object):
    def __init__(self, values=[]):
        self.h = None
        for value in values:
            self.push(value)

    def __len__(self):
        length = 0
        node = self.h
        while node:
            length += 1
            node = node.next()
        return length

    def __iter__(self):
        values = []
        node = self.h
        while node:
            values.append(node.v)
            node = node.next()
        return iter(values)

    def head(self):
        if self.h is None:
            raise EmptyListException('Empty list')

        return self.h

    def push(self, value):
        node = Node(value)
        node.n = self.h
        self.h = node

    def pop(self):
        result = self.head().v
        self.h = self.h.n
        return result

    def reversed(self):
        result = LinkedList()
        node = self.h
        while node:
            result.push(node.v)
            node = node.next()
        return result


class EmptyListException(Exception):
    pass

File: java/simple-linked-list/src/main/java/SimpleLinkedList.java
import java.lang.reflect.Array;
import java.util.NoSuchElementException;

public class SimpleLinkedList<T> {
  Node<T> head;

  SimpleLinkedList() {}

  SimpleLinkedList(T[] values) {
    for (int i = values.length - 1; i >= 0; --i) {
      push(values[i]);
    }
  }

  int size() {
    int result = 0;
    for (Node<T> node = head; node != null; node = node.next) {
      ++result;
    }

    return result;
  }

  void push(T value) {
    Node<T> node = new Node<>(value);
    node.next = head;
    head = node;
  }

  T pop() {
    if (head == null) {
      throw new NoSuchElementException();
    }

    T result = head.value;
    head = head.next;

    return result;
  }

  void reverse() {
    SimpleLinkedList<T> reversed = new SimpleLinkedList<>();
    while (head != null) {
      reversed.push(pop());
    }
    head = reversed.head;
  }

  @SuppressWarnings("unchecked")
  <E> E[] asArray(Class<E> c) {
    E[] result = (E[]) Array.newInstance(c, size());
    Node<T> node = head;
    for (int i = 0; i < result.length; ++i) {
      result[i] = (E) node.value;
      node = node.next;
    }

    return result;
  }
}

class Node<T> {
  T value;
  Node<T> next;

  Node(T value) {
    this.value = value;
  }
}
